[LLVMCPU][ArmSME] Add `2d-scalable-to-1d-scalable` pass

MacDue opened the pull request:
Currently, IREE requires `lowering_config`s to be propagated to all compute ops within a dispatch region (this was recently implemented for scalable vectors in #16350). This can be problematic for SME, as it does not make sense to use 2D-scalability for all ops within a dispatch region.  ArmSME only supports 2D scalable outer products, so if it's not an outer product, we can only scalably vectorize in one dimension.

The solution here is to add a new pass that runs just before vectorization, that drops unsupported scalable tile/vector sizes, producing loops of ops that will only be vectorized scalably in one dimension. This allows earlier passes like `tile-and-fuse` to still function correctly.

Take this simple example:

```mlir
// Lowering configs propagated (from matmul):
linalg.fill {lowering_config = [[4], [4]]
linalg.matmul {lowering_config = [[4], [4], 1]
linalg.generic {lowering_config = [[4], [4]]
```
Here the `linalg.generic` cannot be vectorized with 2D scalable vectors.

After `tile-and-fuse` (which requires consistent lowering configs):
```mlir
scf.for i in range(0, 1000) step 4 x vscale {
  scf.for j in range(0, 2000) step 4 x vscale {
    linalg.fill {lowering_config = [[4], [4]]
    for k in range(0, 100) step 1 {
      linalg.matmul {lowering_config = [[4], [4], 1]
    }
    // 2D scalable vectorization unsupported here:
    linalg.generic {lowering_config = [[4], [4]]
  }
}
```

Unsupported scalability removed (by the pass proposed here):
```mlir
scf.for i in range(0, 1000) step 4 x vscale {
  scf.for j in range(0, 2000) step 4 x vscale {
    linalg.fill {lowering_config = [[4], [4]]
    for k in range(0, 100) step 1 {
      linalg.matmul {lowering_config = [[4], [4], 1]
    }
    // Insert a new loop:
    for n in range(0, 4 x vscale) step 4 {
       // Drop a scalable dim:
       linalg.generic {lowering_config = [4, [4]]
    }
  }
}
```

This can now be vectorized and lowered successfully, which produces a dispatch that mixes SME and SVE.

MacDue committed:
[LLVMCPU][ArmSME] Add `unsupported-scalability-to-loops` pass

Currently, IREE requires `lowering_config`s to be propagated to all
compute ops within a dispatch region (this was recently implemented for
scalable vectors in #16350). This can be problematic for SME, as it does
not make sense to use 2D-scalability for all ops within a dispatch
region. SME can only really do 2D-scalable outer products (everything
else is still 1D, like SVE).

The solution here is to add a new pass that runs just before
vectorization, that drops unsupported scalable tile/vector sizes,
producing loops of ops that will only be vectorized scalably in one
dimension. This allows earlier passes like `tile-and-fuse` to still
function correctly.

Take this simple example:

```mlir
// Lowering configs propagated (from matmul):
linalg.fill {lowering_config = [[4], [4]]
linalg.matmul {lowering_config = [[4], [4], 1]
linalg.generic {lowering_config = [[4], [4]]
```
Here the `linalg.generic` cannot be vectorized with 2D scalable vectors.

After `tile-and-fuse` (which requires consistent lowering configs):
```mlir
scf.for i in range(0, 1000) step 4 x vscale {
  scf.for j in range(0, 2000) step 4 x vscale {
    linalg.fill {lowering_config = [[4], [4]]
    for k in range(0, 100) step 1 {
      linalg.matmul {lowering_config = [[4], [4], 1]
    }
    // 2D scalable vectorization unsupported here:
    linalg.generic {lowering_config = [[4], [4]]
  }
}
```

Unsupported scalability removed:
```
scf.for i in range(0, 1000) step 4 x vscale {
  scf.for j in range(0, 2000) step 4 x vscale {
    linalg.fill {lowering_config = [[4], [4]]
    for k in range(0, 100) step 1 {
      linalg.matmul {lowering_config = [[4], [4], 1]
    }
    // Insert a new loop:
    for n in range(0, 4 x vscale) step 4 {
       // Drop a scalable dim:
       linalg.generic {lowering_config = [4, [4]]
    }
  }
}
```

This can now be vectorized and lowered successfully, which produces a
dispatch that mixes SME and SVE.

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUUnsupportedScalabilityToLoops.cpp:
```diff
@@ -0,0 +1,138 @@
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-llvmcpu-unsupported-scalability-to-loops"
+#define VEC_DBGS() (llvm::dbgs() << '[' << DEBUG_TYPE << "] ")
```
Not used

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUUnsupportedScalabilityToLoops.cpp:
```diff
@@ -0,0 +1,138 @@
+  LogicalResult matchAndRewrite(TilingInterface op,
+                                PatternRewriter &rewriter) const override {
+    // Note: This rewrite is currently only required for ArmSME (which is the
+    // only target that currently has some concept of 2D scalability).
```
IMHO this comment fits better in [compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.td](https://github.com/openxla/iree/pull/16712/files#diff-1e3b594a934d4a19e8f736d21de5e0a3f2e121b629a06a26247265c6d30089e4) - it's a general comment on the applicability of this pass. Definitely worth adding a note like this at the top of this file.

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/test/unsupported_scalability_to_loops.mlir:
```diff
@@ -0,0 +1,75 @@
+//      CHECK:         linalg.generic
+//      CHECK:       }
+//      CHECK:     }
+//      CHECK:   }
```
Would the lowering configs also get updated? Could you check that as well?

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUUnsupportedScalabilityToLoops.cpp:
```diff
@@ -0,0 +1,138 @@
+    auto tileSizes = loweringConfigAttr.getTileSizeVals();
+    auto scalableFlags = loweringConfigAttr.getScalableTileFlagVals();
+
+    // Drop scalable dimensions from leading tiling levels first (and leading
```
```suggestion
    // 1. Drop scalable dimensions from leading tiling levels first (and leading
```

This is a bit complex, hence suggesting to clearly mark separate steps.

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUUnsupportedScalabilityToLoops.cpp:
```diff
@@ -0,0 +1,138 @@
+      }
+    }
+
+    // Re-tile the operation with some scalability dropped. This introduces
```
```suggestion
    // 2. Re-tile the operation with some scalability dropped. This introduces
```

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUUnsupportedScalabilityToLoops.cpp:
```diff
@@ -0,0 +1,138 @@
+    if (failed(tilingResult))
+      return failure();
+
+    // Update the lowering config of the new tiled operations.
```
```suggestion
    // 3. Update the lowering config of the new tiled operations.
```

banach-space reviewed (commented):
Makes sense, thanks for implementing this!

I think that `LLVMCPUUnsupportedScalabilityToLoopsPass` is a bit too specific (i.e. the name). In particular, the fact that the pass will only process operations that don't support "scalability" in 2 dimensions is, IMHO, an implementation detail. Instead, I would rename as:

*  `LLVMCPU2DScalableTo1DScalable`

You can add comments in the code explaining that:

1. For SME, 2 dimensions are marked as scalable for `linalg.matmul`.
2. Scalable flags are propagated as part of the generic lowering path.
3. This needs to be refined when dealing with e.g. `linalg.generic` (and other Ops) that can only have 1 dim scalable.
4. We will likely relax this in the future as we learn how to better use "scalability" in 2D.

You have already covered this in your very detailed summary (thanks!), so it should just be a matter copy & paste.

MacDue committed:
Fixups (renaming + comments)

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPU2DScalableTo1DScalable.cpp:
```diff
@@ -0,0 +1,193 @@
+      patterns.getContext(), assumeArmSME);
+
+  if (mlir::failed(mlir::applyPatternsAndFoldGreedily(getOperation(),
+                                                      std::move(patterns)))) {
```
errant extra line?

MacDue replied:
This is just the result of `clang-format` line wrapping :pensive:

ScottTodd replied:
One thing you can do is drop the `mlir::` since you're already in the `mlir::iree_compiler` namespace:
```c++
  if (failed(
          applyPatternsAndFoldGreedily(getOperation(), std::move(patterns)))) {
    signalPassFailure();
  }
```

I don't actually see any uses of `mlir::applyPatternsAndFoldGreedily` or `mlir::failed(` in IREE, all omit that prefix.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPU2DScalableTo1DScalable.cpp:
```diff
@@ -0,0 +1,193 @@
+  patterns.add<DropUnsupportedScalableDimsFromTilingInterfaceOps>(
+      patterns.getContext(), assumeArmSME);
+
+  if (mlir::failed(mlir::applyPatternsAndFoldGreedily(getOperation(),
```
Doing tiling with patterns has been tricky. It might be easier to just walk the IR and get the ops you want to transform and apply the tiling transformation to it directly (i..e not use the pattern rewriter mechanism)

MacDue replied:
What's the issue? I didn't notice a problem while working on this :sweat_smile:

hanhanW replied:
Pattern rewriting is not a good fit for tiling. Learning from the past development, we usually need a filter/attribute to represent the state of the op. E.g., which op should be tiled, and which op is already tiled. It introduces the "state" semantics in the IR. Otherwise, it falls in an infinite pattern rewriting loop. If we take a close look at what should be happening, it is always

1. Collect the target operations.
2. Apply the tiling.
3. End.

In your use case, it is fine because the matcher is "whether the op has a lowering_config or not". IMO, we want to get rid of filter-based tiling patterns. Having a consistent way to apply tiling in IREE could prevent such patterns created again. So I'd vote to not use patterns here.

hanhanW replied:
> it is fine because the matcher is "whether the op has a lowering_config or not"

Sorry, I misread the code. So you're recursively to undo scalable tiling? I think in this case, we can use the `getVectorTileSizes` method from `TilingConfig`. Because scalable flag is only meaningful to vector level of tiling. (we should probably add it to the pipeline verifier). And we can easily extend this from `2D->1D` to `nD->1D`.

side note: I'm thinking if we can carry such tiling information to lowering_config, so we don't need to reset lowering_config. Updating configs during lowering pipeline is not a good sign because users do not expect that. One of examples that I can think of is about tuning. In tuning, we can preset configs, compile and benchmark. This could lead users to a weird state. Because the codegen behavior could be different from what users expect.  I don't have a solution yet, though.

MacDue replied:
I'm aware of `getVectorTileSizes()` but I didn't find it to be very helpful here. The vector sizes can be split between the parallel, reduction, and inner parallel tiling levels, and that information is not returned from that function. So since I want to update the lowering config attribute, I found it easier just to look at it directly.

MacDue replied:
> And we can easily extend this from 2D->1D to nD->1D

There's already a comment pointing out this can easily be changed to `nD to mD` (where m < n) :)

hanhanW replied:
I think `getVectorTileSizes` returns a joint list from parallel, reduction and inner parallel tile sizes. That gives you all you need right? E.g., the lowering_config on a matmul op could be `[[d_x, d_y, 0], [[4], [4], 0], [0, 0, 1], [0, 0, 0]]`; the `getVectorTileSizes()` returns `[[4], [4], 1]` in this case.

MacDue replied:
I'm not sure I follow. I want to update the `lowering_config`, having the joint list i.e. `[[4], [4], 1]` does not help because I have to map that back to the `lowering_config` tiling levels (which is not a 1-to-1 mapping, there's multiple lowering configs which could result in the same `getVectorTileSizes()`).

hanhanW replied:
Sorry, I missed the update part. In this case, I think we want to add a method to update vector tile sizes in TilingConfig. The main concern is about updating lowering_config. Updating lowering_config after KernelDispatch stage is a red flag to me. If we can't get rid of it, let's at least access and update it through a defined class/struct (which is TilingConfig in this case).

MacDue replied:
The `TilingConfig` does not provide any way to update tile sizes, it simply provides methods for reading the `lowering_config`.

hanhanW replied:
I meant it is a missing feature if we want to update lowering_config. We need to implement it. I can try to implement it if you're more comfortable about it.

MacDue replied:
I can implement it, but it'll have to re-create the `lowering_config` attribute (as I believe the actual attribute data is immutable).

MacDue replied:
i.e. Something like:
```c++
LoweringConfigAttr TilingConfig::withNewVectorSizes(ArrayRef<int64_t> sizes, ArrayRef<bool> scalableFlags);
```

hanhanW replied:
This sounds good to me, thanks!

MacDue replied:
Updated to avoid using patterns, and use the suggested `withNewVectorSizes()` helper.

MaheshRavishankar reviewed (changes requested):
high level looks OK to me (just have one comment), but I would expect this is something that could happen much later in the pipeline? I see why you have it here for now, but just looking in from outside this seems more like a "operation legalization" thing.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:
```diff
@@ -393,6 +393,11 @@ void addMultiTilingExpertPassPipeline(OpPassManager &passManager,
     }
   }
 
+  if (pipelineOpt.enableAArch64SSVE) {
```
I missed it in the previous review. Should this happen after peeling? i.e., should it happen right before vectorization?

MacDue replied:
Good point :+1:

MacDue committed:
Introduce `TilingConfig::withNewVectorSizes(sizes, scalableFlags)` helper

This provides a safe method of updating the vector sizes of a
`lowering_config`.

MacDue committed:
Address review comments in `2d-scalable-to-1d-scalable` pass

- Use worklist rather than pattern rewriters
- Make use of `tilingConfig.withNewVectorSizes()` helper

MacDue committed:
Fixups

MacDue committed:
Refine test

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Common/TileSizeSelection.h:
```diff
@@ -113,6 +113,12 @@ class TilingConfig {
+  /// dimensions, set to `sizes`, and the corresponding scalability set to
+  /// `scalableFlags`.
+  IREE::Codegen::LoweringConfigAttr
+  withNewVectorSizes(ArrayRef<int64_t> sizes, ArrayRef<bool> scalableFlags);
```
Nit: rename to `getLoweringConfigWithNewVectorSizes` .. wider names are OK by me.

MaheshRavishankar reviewed (approved):
Ok, looks good to me. THanks!

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPU2DScalableTo1DScalable.cpp:
```diff
@@ -0,0 +1,175 @@
+    // 2. Re-tile the operation with some scalability dropped. This introduces
+    // loops for previously scalable vector/tile sizes.
+    scf::SCFTilingOptions options{};
+    setSCFTileSizes(options, tilingOp, loopTileSizes, {});
```
nit: add function argument comment for better readability

```suggestion
    setSCFTileSizes(options, tilingOp, loopTileSizes, /*tileScalableFlags=*/{});
```

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPU2DScalableTo1DScalable.cpp:
```diff
@@ -0,0 +1,175 @@
+
+    // 2. Re-tile the operation with some scalability dropped. This introduces
+    // loops for previously scalable vector/tile sizes.
+    scf::SCFTilingOptions options{};
```
I think we don't need `{}` to initialize the struct.

```suggestion
    scf::SCFTilingOptions options;
```

hanhanW reviewed (commented):
LGTM, just two nits!

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPU2DScalableTo1DScalable.cpp:
```diff
@@ -0,0 +1,175 @@
+/// ops within a dispatch region. This can be problematic for SME, as it does
+/// not make sense to use 2D-scalability for all ops within a dispatch region.
+/// The Arm SME target only supports 2D scalable outer products, so if it's not
+/// an outer product, we can only scalably vectorize in one dimension.
```
[nit] I would avoid terms like "make sense" in docs - they tend to be quite subjective 😅 Instead, lets keep this short and focus on the ISA aspects that drive the need for this.

```suggestion
/// ops within a dispatch region. This can be problematic for SME which only supports
// 2D scalable outer product operations - if an operation cannot be lowered to an
/// outer product, we can only scalably vectorize in one dimension.
```

Also, is it really "outer-product" that's key here or is it the fact that we need something that works with `ZA`? It just happens that today we only have `linalg.matmul` + `linalg.fill` that we lower to outer-products. But this list is likely to expand.

MacDue replied:
I think having outer products is key. There's other operations that can use ZA, but AFAIK only the outer product instructions are scalable in two dimensions. Other ZA instructions deal with vector groups, which would only be scalable in one dimension.

banach-space commented on compiler/src/iree/compiler/Codegen/Common/TileSizeSelection.cpp:
```diff
@@ -92,6 +92,76 @@ SizesAndScalableFlags TilingConfig::getVectorTileSizes() {
+    auto level = tilingLevels[tilingLevelIndex];
+    return level.getSizes()[dimPos] != 0;
+  };
+  auto tilingLevelForDimPos = [&](unsigned dimPos) -> unsigned {
```
[nit] `getTilingLevelForDimPos`? Also, this feels like a useful "utility" function - why not move it to TileSizeSelection.h?

MacDue replied:
Done, and I've also made use of it for `getVectorTileSizes()` now (which I think simplifies it quite a bit :))

banach-space commented on compiler/src/iree/compiler/Codegen/Common/TileSizeSelection.cpp:
```diff
@@ -92,6 +92,76 @@ SizesAndScalableFlags TilingConfig::getVectorTileSizes() {
+                                           unsigned dimPos) mutable {
+    auto level = tilingLevels[tilingLevelIndex];
+    return level.getSizes()[dimPos] != 0;
+  };
```
```suggestion
  bool isDimSizeNonZeroForTilingLevel = [tilingLevels](unsigned tilingLevelIndex,
                                           unsigned dimPos) mutable {
    auto level = tilingLevels[tilingLevelIndex];
    return level.getSizes()[dimPos] != 0;
  };
```

It's a lambda rather than a function, but still feels that following this makes sense: https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly

banach-space commented on compiler/src/iree/compiler/Codegen/Common/TileSizeSelection.cpp:
```diff
@@ -92,6 +92,76 @@ SizesAndScalableFlags TilingConfig::getVectorTileSizes() {
+
+  MLIRContext *context = loweringConfig.getContext();
+  SmallVector<IREE::Codegen::LoweringConfigTilingLevelAttr>
+      updatedTilingLevelsList(tilingLevels.begin(), tilingLevels.end());
```
[nit] "new" rather than "updated"? To match with the function name - this way it would be clearer that this where the "new" stuff is being created.

banach-space reviewed (approved):
Thanks Ben, makes sense! A few final nis/suggestions, but all in all looks good :)

MacDue committed:
Review fixups and slight refactoring

Addresses comments & moves some common code between `getVectorTileSizes()`
and `getLoweringConfigWithNewVectorSizes()` to shared helpers.
